File: LeetCode/POTDs/unique paths3 .cpp
class Solution {
public:
    int uniquePathsIII(vector<vector<int>>& grid) {
        int t_x,t_y,noOfz=0,row=grid.size(),col=grid[0].size();
        for(int i=0;i<row;i++){
            for(int j=0;j<col;j++){
                if (grid[i][j]==0){
                noOfz+=1;
                }
                else if (grid[i][j]==1){
                    t_x=i;
                    t_y=j;
                }
            }
        }
        int ans=0;
        int z=0;
        dfs(grid,t_x,t_y,row,col,z,noOfz,ans);
        return ans;
    }
    void dfs(vector<vector<int>>& grid,int x,int y,int r,int c,int z,int& noOfz,int& ans){
        if (x<0 || x>=r || y<0 || y>=c || grid[x][y]==-1){
            return ;
        }
        if(grid[x][y]==2){
            if (z-1==noOfz)
            ans+=1;
            return ;
        }
        grid[x][y]=-1;
        vector<pair<int,int>>dir={{0,1},{1,0},{-1,0},{0,-1}};
        for(pair<int,int>p:dir){
            int i=x+p.first;
            int j=y+p.second;
            if (x>=0 && x<r && y>=0 && y<c){
                dfs(grid,i,j,r,c,z+1,noOfz,ans);
            }
        }
       
        grid[x][y]=0;
        
        
    }
};

File: LeetCode/POTDs/unique paths 3.py
class Solution:
    def dfs(self,grid,x,y,r,c,z):
        if x<0 or x>=r or y<0 or y>=c or grid[x][y]==-1:
            return 0
        if grid[x][y]==2:
            if z==-1:
                return 1 
            else:
                return 0
        grid[x][y]=-1
        z-=1
        tot_p=self.dfs(grid,x+1,y,r,c,z)+self.dfs(grid,x-1,y,r,c,z)+self.dfs(grid,x,y+1,r,c,z)+self.dfs(grid,x,y-1,r,c,z)

        grid[x][y]=0
        z+=1

        return tot_p
        
    def uniquePathsIII(self, grid: List[List[int]]) -> int:
        row=len(grid)
        col=len(grid[0])
        noOfZeros=0
        for i in range(row):
            for j in range(col):
                if grid[i][j]==0: noOfZeros+=1
                elif grid[i][j]==1:
                    target_x,target_y=i,j
        return self.dfs(grid,target_x,target_y,row,col,noOfZeros)
                
